test/test_transaction.py

```diff
@@ -23,26 +27,34 @@
 class TestTransaction:  # pylint: disable=too-many-public-methods
     """Unittest for the pymodbus.transaction module."""
 
-    client = None
-    decoder = None
-    _tcp = None
-    _tls = None
-    _rtu = None
-    _ascii = None
-    _manager = None
-    _tm = None
+    client: ModbusBaseSyncClient
+    decoder: ServerDecoder
+    _tcp: ModbusSocketFramer
+    _tls: ModbusTlsFramer
+    _rtu: ModbusRtuFramer
+    _ascii: ModbusAsciiFramer
+    _manager: SyncModbusTransactionManager
 
     # ----------------------------------------------------------------------- #
     # Test Construction
     # ----------------------------------------------------------------------- #
     def setup_method(self):
         """Set up the test environment."""
-        self.client = None
+        self.client = mock.create_autospec(ModbusBaseSyncClient)
+        self.client.framer = mock.create_autospec(ModbusFramer)
         self.decoder = ServerDecoder()
-        self._tcp = ModbusSocketFramer(decoder=self.decoder, client=None)
-        self._tls = ModbusTlsFramer(decoder=self.decoder, client=None)
-        self._rtu = ModbusRtuFramer(decoder=self.decoder, client=None)
-        self._ascii = ModbusAsciiFramer(decoder=self.decoder, client=None)
+        self._tcp = ModbusSocketFramer(
+            decoder=self.decoder, client=mock.create_autospec(ModbusTcpClient)
```

Comment:
We are not using the client, and if the production code start using the client, having it as None ensures a failure. Your change removes that detection.

Reply:
Ah, so looks like mypy wasn't flagging this as a type issue but pycharm was, weird. it's getting picked up as an issue since this type hint does not have `None` as an option, but adding `None` causes other type check issues. I'll look at that later reverted for now.

https://github.com/pymodbus-dev/pymodbus/blob/385bbc5ef6d2189f7426961e7461e6330ff0208c/pymodbus/framer/old_framer_base.py#L36